api/admin/controller/sign_in.py

```diff
@@ -67,6 +75,99 @@ class SignInController(AdminController):
+
+        result = flask.request.form.get("result")
+        if result != "success":
+            logging.error("Ekirjasto authentication failed, result = %s", result)
+            return self.error_response(EKIRJASTO_REMOTE_AUTHENTICATION_FAILED)
+
+        ekirjasto_token = flask.request.form.get("token")
+
+        user_info = auth.ekirjasto_authenticate(ekirjasto_token)
+        if isinstance(user_info, ProblemDetail):
+            return user_info
+
+        circulation_role = self._to_circulation_role(user_info.role)
+        if not circulation_role:
+            return self.error_response(ADMIN_NOT_AUTHORIZED)
+
+        try:
+            credentials = get_one(self._db, AdminCredential, external_id=user_info.sub)
+            if credentials:
+                admin = credentials.admin
+            else:
+                admin = self._create_admin_with_external_credentials(user_info)
+
+            self._update_roles_if_changed(admin, circulation_role)
+        except Exception as e:
+            logging.exception("Internal error during signup")
+            self._db.rollback()
+            return EKIRJASTO_REMOTE_AUTHENTICATION_FAILED
+
+        self._setup_admin_flask_session(admin, auth, ekirjasto_token)
+
+        redirect_uri = flask.request.args.get("redirect_uri", "/admin/web")
+        return SanitizedRedirections.redirect(redirect_uri)
+
+    def _create_admin_with_external_credentials(self, user_info: EkirjastoUserInfo):
+        admin, _ = get_one_or_create(
+            self._db,
+            Admin,
+            email=user_info.sub,
+        )
+        get_one_or_create(
+            self._db,
+            AdminCredential,
+            external_id=user_info.sub,
+            admin_id=admin.id,
+        )
+        return admin
+
+    @staticmethod
+    def _update_roles_if_changed(admin: Admin, new_role: str):
+        existing_roles = [(role.role, role.library) for role in admin.roles]
+        if [(new_role, None)] != existing_roles:
+            for role in admin.roles:
+                admin.remove_role(role.role, role.library)
+            admin.add_role(new_role)
+
+    @staticmethod
+    def _setup_admin_flask_session(
+        admin: Admin,
+        auth: EkirjastoAdminAuthenticationProvider,
+        ekirjasto_token: str,
+    ):
+        # Set up the admin's flask session.
+        flask.session["admin_email"] = admin.email
+        flask.session["auth_type"] = auth.NAME
+
+        # This one is extra compared to password auth provider
+        flask.session["ekirjasto_token"] = ekirjasto_token
+
+        # A permanent session expires after a fixed time, rather than
+        # when the user closes the browser.
+        flask.session.permanent = True
+
+    @staticmethod
+    def _to_circulation_role(ekirjasto_role: str) -> str | None:
+        if ekirjasto_role == "orgadmin":
+            return AdminRole.SYSTEM_ADMIN
+        elif ekirjasto_role == "admin":
+            return AdminRole.SITEWIDE_LIBRARY_MANAGER
```

Comment:
The local library manager can access the collection menu (aka left menu) with localized statistics and localized custom lanes.